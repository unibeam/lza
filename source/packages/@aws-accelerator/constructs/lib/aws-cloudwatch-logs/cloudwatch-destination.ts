/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * Construction properties for CloudWatch Destination for Kinesis Stream.
 */
export interface CloudWatchDestinationProps {
  /**
   *
   * KMS key to encrypt the Kinesis Data Stream
   */
  kinesisKmsKey: cdk.aws_kms.IKey;
  /**
   *
   * Kinesis Data Stream for CloudWatch logs
   */
  kinesisStream: cdk.aws_kinesis.IStream;
  /**
   * Organization ID to restrict the usage within specific org
   */
  organizationId?: string;
  /**
   * Partition to determine the IAM condition.
   */
  partition: string;
  /**
   * Account IDs for the IAM condition.
   */
  accountIds?: string[];
  /**
   * Accelerator Prefix defaults to 'AWSAccelerator'.
   */
  acceleratorPrefix: string;
  /**
   * Use existing IAM roles for deployment.
   */
  useExistingRoles: boolean;
}
/**
 * Class to configure CloudWatch Destination on logs receiving account
 */
export class CloudWatchDestination extends Construct {
  constructor(scope: Construct, id: string, props: CloudWatchDestinationProps) {
    super(scope, id);

    let principalOrgIdCondition: object | undefined = undefined;
    let accountPrincipals: object | cdk.aws_iam.IPrincipal;

    if (props.partition === 'aws-cn' || !props.organizationId) {
      // Only principal block with list of account id is supported.
      accountPrincipals = {
        AWS: props.accountIds,
      };
    } else {
      principalOrgIdCondition = {
        StringEquals: {
          'aws:PrincipalOrgID': props.organizationId,
        },
      };
      accountPrincipals = new cdk.aws_iam.AnyPrincipal();
    }

    const logDestinationPolicy = JSON.stringify({
      Version: '2012-10-17',
      Statement: [
        {
          Effect: 'Allow',
          Principal: accountPrincipals,
          Action: 'logs:PutSubscriptionFilter',
          Resource: `arn:${cdk.Stack.of(this).partition}:logs:${cdk.Stack.of(this).region}:${
            cdk.Stack.of(this).account
          }:destination:${props.acceleratorPrefix}CloudWatchToS3`,
          Condition: principalOrgIdCondition,
        },
      ],
    });

    new cdk.aws_logs.CfnDestination(this, 'Resource', {
      roleArn: this.createKinesisRole(
        props.kinesisStream.streamArn,
        props.kinesisKmsKey.keyArn,
        props.useExistingRoles,
        props.acceleratorPrefix,
      ),
      targetArn: props.kinesisStream.streamArn,
      destinationName: `${props.acceleratorPrefix}CloudWatchToS3`,
      destinationPolicy: logDestinationPolicy,
    });
  }

  private createKinesisRole(
    kinesisStreamArn: string,
    kinesisKeyArn: string,
    useExistingRoles: boolean,
    acceleratorPrefix: string,
  ) {
    if (useExistingRoles) {
      return `arn:${cdk.Stack.of(this).partition}:iam::${
        cdk.Stack.of(this).account
      }:role/${acceleratorPrefix}LogReplicationRole-${cdk.Stack.of(this).region}`;
    }
    //Create policy for access to Kinesis stream
    const kinesisStreamAccess = new cdk.aws_iam.PolicyDocument({
      statements: [
        new cdk.aws_iam.PolicyStatement({
          actions: ['kinesis:ListShards', 'kinesis:PutRecord', 'kinesis:PutRecords'],
          resources: [kinesisStreamArn],
        }),
      ],
    });
    const kmsKeyAccess = new cdk.aws_iam.PolicyDocument({
      statements: [
        new cdk.aws_iam.PolicyStatement({
          actions: [
            'kms:Decrypt',
            'kms:Encrypt',
            'kms:GenerateDataKey',
            'kms:ReEncryptTo',
            'kms:GenerateDataKeyWithoutPlaintext',
            'kms:GenerateDataKeyPairWithoutPlaintext',
            'kms:GenerateDataKeyPair',
            'kms:ReEncryptFrom',
          ],
          resources: [kinesisKeyArn],
        }),
      ],
    });
    // Create a role for CloudWatch Logs destination
    const logsKinesisRole = new cdk.aws_iam.Role(this, 'LogsKinesisRole', {
      assumedBy: new cdk.aws_iam.ServicePrincipal(`logs.${cdk.Stack.of(this).region}.amazonaws.com`),
      // this needs to be inline to ensure role is created with proper access
      // if not, CloudWatch destination creation fails with no permission to access Kinesis or KMS (generateDataKey access error)
      inlinePolicies: {
        KinesisAccess: kinesisStreamAccess,
        KmsAccess: kmsKeyAccess,
      },
    });
    return logsKinesisRole.roleArn;
  }
}
