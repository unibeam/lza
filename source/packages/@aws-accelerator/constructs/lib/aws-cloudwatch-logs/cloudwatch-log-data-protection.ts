/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { LzaCustomResource } from '../lza-custom-resource';
import * as path from 'path';

/**
 * Construction properties for CloudWatch CloudWatchLogDataProtection.
 */
export interface CloudWatchLogDataProtectionProps {
  /**
   *
   * Central logs bucket name
   */
  centralLogBucketName: string;
  /**
   *
   * Data protection identifier names
   */
  identifierNames: string[];
  /**
   * Indicates whether existing CloudWatch Log data protection policy configuration can be overwritten.
   */
  overrideExisting: boolean;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly customResourceLambdaEnvironmentEncryptionKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly customResourceLambdaCloudWatchLogKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly customResourceLambdaLogRetentionInDays: number;
}
/**
 * Class to configure CloudWatch log data protection
 */
export class CloudWatchLogDataProtection extends Construct {
  constructor(scope: Construct, id: string, props: CloudWatchLogDataProtectionProps) {
    super(scope, id);

    const resourceName = 'CloudWatchDataProtection';

    new LzaCustomResource(this, resourceName, {
      resource: {
        name: resourceName,
        parentId: id,
        properties: [
          {
            centralLogBucketName: props.centralLogBucketName,
            identifierNames: props.identifierNames,
            partition: cdk.Stack.of(this).partition,
            overrideExisting: props.overrideExisting,
          },
        ],
      },
      lambda: {
        assetPath: path.join(__dirname, 'put-account-policy/dist'),
        environmentEncryptionKmsKey: props.customResourceLambdaEnvironmentEncryptionKmsKey,
        cloudWatchLogKmsKey: props.customResourceLambdaCloudWatchLogKmsKey,
        cloudWatchLogRetentionInDays: props.customResourceLambdaLogRetentionInDays,
        timeOut: cdk.Duration.minutes(15),
        roleInitialPolicy: [
          new cdk.aws_iam.PolicyStatement({
            sid: 'CloudWatchAccess',
            effect: cdk.aws_iam.Effect.ALLOW,
            actions: [
              'logs:DeleteAccountPolicy',
              'logs:DescribeAccountPolicies',
              'logs:PutAccountPolicy',
              'logs:PutDataProtectionPolicy',
              'logs:DeleteDataProtectionPolicy',
              'logs:CreateLogDelivery',
            ],
            resources: ['*'],
          }),
          new cdk.aws_iam.PolicyStatement({
            sid: 'S3Access',
            effect: cdk.aws_iam.Effect.ALLOW,
            actions: ['s3:GetBucketPolicy', 's3:PutBucketPolicy'],
            resources: [`arn:${cdk.Stack.of(this).partition}:s3:::${props.centralLogBucketName}`],
          }),
        ],
      },
    });
  }
}
