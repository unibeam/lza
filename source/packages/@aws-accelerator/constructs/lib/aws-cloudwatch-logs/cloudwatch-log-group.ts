/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as cdk from 'aws-cdk-lib';
import * as path from 'path';
import { pascalCase } from 'change-case';
import { Construct } from 'constructs';

/**
 * Construction properties for CloudWatch LogGroups.
 */

export type LogGroupRetention =
  | 1
  | 3
  | 5
  | 7
  | 14
  | 30
  | 60
  | 90
  | 120
  | 150
  | 180
  | 365
  | 400
  | 545
  | 731
  | 1096
  | 1827
  | 2192
  | 2557
  | 2922
  | 3288
  | 3653;

export interface CloudWatchLogGroupsProps {
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly customLambdaLogKmsKey?: cdk.aws_kms.IKey;
  /**
   * How long, in days, the log contents for the Lambda function
   * will be retained.
   */
  readonly customLambdaLogRetention: number;
  /**
   * How long, in days, the log contents will be retained.
   *
   * To retain all logs, set this value to undefined.
   *
   */
  readonly logRetentionInDays: LogGroupRetention | number;
  /**
   * KMS Key Arn to encrypt CloudWatch Logs Group at rest.
   */
  readonly keyArn?: string;
  /**
   *
   * Name of the CloudWatch Logs Group
   */
  readonly logGroupName?: string;
  /**
   * For cross-account log groups, the owning account ID
   */
  readonly owningAccountId?: string;
  /**
   * For cross-region log groups, the owning region
   */
  readonly owningRegion?: string;
  /**
   * For cross-account log groups, the IAM role name to assume
   */
  readonly roleName?: string;
  /**
   *
   * Determine termination policy on CloudWatch Logs Group
   */
  readonly terminationProtected?: boolean;
}

interface ILogGroup {
  /**
   * The name of the log group
   */
  readonly logGroupName: string;
  /**
   * The ARN of the log group
   */
  readonly logGroupArn: string;
}

/**
 * Class to configure CloudWatch Log Groups
 */
export class CloudWatchLogGroups extends cdk.Resource implements ILogGroup {
  public readonly logGroupName: string;
  public readonly logGroupArn: string;

  constructor(scope: Construct, id: string, props: CloudWatchLogGroupsProps) {
    super(scope, id);
    const CLOUD_WATCH_LOG_GROUPS = 'Custom::CreateLogGroup';
    this.logGroupName = props.logGroupName ?? cdk.Names.uniqueResourceName(this, { separator: '-' });

    //
    // Function definition for the custom resource
    //
    const policyStatements = props.owningAccountId
      ? [
          {
            Effect: 'Allow',
            Action: ['sts:AssumeRole'],
            Resource: `arn:${this.stack.partition}:iam::*:role/${props.roleName}`,
          },
        ]
      : [
          {
            Effect: 'Allow',
            Action: ['logs:CreateLogGroup', 'logs:DeleteLogGroup', 'logs:PutRetentionPolicy'],
            Resource: `arn:${this.stack.partition}:logs:${props.owningRegion ?? this.stack.region}:${
              this.stack.account
            }:log-group:${this.logGroupName}:*`,
          },
          {
            Effect: 'Allow',
            Action: ['kms:DescribeKey', 'kms:ListKeys', 'logs:AssociateKmsKey', 'logs:DescribeLogGroups'],
            Resource: '*',
          },
        ];

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(
      this,
      pascalCase(`${this.logGroupName}-${CLOUD_WATCH_LOG_GROUPS}`),
      {
        codeDirectory: path.join(__dirname, 'create-log-groups/dist'),
        runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
        policyStatements,
      },
    );

    const resource = new cdk.CustomResource(this, 'CloudWatchLogsResource', {
      resourceType: CLOUD_WATCH_LOG_GROUPS,
      serviceToken: provider.serviceToken,
      properties: {
        logGroupName: this.logGroupName,
        retention: props.logRetentionInDays,
        keyArn: props.keyArn,
        owningAccountId: props.owningAccountId,
        owningRegion: props.owningRegion,
        roleName: props.roleName,
        terminationProtected: props.terminationProtected,
      },
    });
    this.logGroupArn = resource.getAttString('LogGroupArn');

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.customLambdaLogRetention,
        encryptionKey: props.customLambdaLogKmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);
  }
}
