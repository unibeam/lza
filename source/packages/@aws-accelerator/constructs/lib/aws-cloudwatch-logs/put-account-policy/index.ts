/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';

import { setRetryStrategy } from '@aws-accelerator/utils/lib/common-functions';

import {
  AccountPolicy,
  CloudWatchLogsClient,
  DeleteAccountPolicyCommand,
  DescribeAccountPoliciesCommand,
  PolicyType,
  PutAccountPolicyCommand,
  Scope,
} from '@aws-sdk/client-cloudwatch-logs';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';

/**
 * create-log-groups - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string;
    }
  | undefined
> {
  const centralLogBucketName: string = event.ResourceProperties['centralLogBucketName'];
  const identifierNames: string[] = event.ResourceProperties['identifierNames'];
  const overrideExisting = event.ResourceProperties['overrideExisting'] === 'true' ? true : false;
  const partition = event.ResourceProperties['partition'];

  const solutionId = process.env['SOLUTION_ID'];

  const client = new CloudWatchLogsClient({
    customUserAgent: solutionId,
    retryStrategy: setRetryStrategy(),
  });

  const existingPolicies = await getExistingPolicies(client);

  let isPolicyExists = false;

  if (existingPolicies.length === 1) {
    isPolicyExists = true;
  }

  let policyName = 'ACCELERATOR_ACCOUNT_DATA_PROTECTION_POLICY';

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      if (isPolicyExists && !overrideExisting) {
        console.warn(
          `Existing policy ${existingPolicies[0]
            .policyName!} found, and override existing flag is set to false, skip update of policy.`,
        );
        return {
          Status: 'SUCCESS',
        };
      }

      if (isPolicyExists) {
        policyName = existingPolicies[0].policyName!;
        console.log(
          `Existing policy ${existingPolicies[0]
            .policyName!} found, and override existing flag is set to true, policy will be updated.`,
        );
      } else {
        console.log(`No existing policy found, policy ${policyName} will be created.`);
      }

      const dataIdentifiers: string[] = identifierNames.map(
        item => `arn:${partition}:dataprotection::${partition}:data-identifier/${item}`,
      );

      const policyDocument = {
        Name: policyName,
        Description: 'Accelerator deployed CloudWatch log data protection policy',
        Version: '2021-06-01',
        Statement: [
          {
            Sid: 'audit-policy',
            DataIdentifier: dataIdentifiers,
            Operation: {
              Audit: {
                FindingsDestination: {
                  S3: {
                    Bucket: centralLogBucketName,
                  },
                },
              },
            },
          },
          {
            Sid: 'redact-policy',
            DataIdentifier: dataIdentifiers,
            Operation: {
              Deidentify: {
                MaskConfig: {},
              },
            },
          },
        ],
      };

      await throttlingBackOff(() =>
        client.send(
          new PutAccountPolicyCommand({
            policyName: policyName,
            policyDocument: JSON.stringify(policyDocument),
            policyType: PolicyType.DATA_PROTECTION_POLICY,
            scope: Scope.ALL,
          }),
        ),
      );

      return {
        Status: 'SUCCESS',
      };

    case 'Delete':
      if (isPolicyExists) {
        // Delete policy deployed by the solution only
        if (existingPolicies[0].policyName! === policyName) {
          console.log(
            `Existing policy ${existingPolicies[0]
              .policyName!} found, which is similar to solution deployed policy, policy will be deleted.`,
          );
          await throttlingBackOff(() =>
            client.send(
              new DeleteAccountPolicyCommand({
                policyName: policyName,
                policyType: PolicyType.DATA_PROTECTION_POLICY,
              }),
            ),
          );
        }
      }

      return {
        Status: 'SUCCESS',
      };
  }

  /**
   * Function to get existing account policy configuration
   * @param client {@link CloudWatchLogsClient}
   * @returns policyConfiguration {@link AccountPolicy}[]
   */
  async function getExistingPolicies(client: CloudWatchLogsClient): Promise<AccountPolicy[]> {
    const response = await throttlingBackOff(() =>
      client.send(
        new DescribeAccountPoliciesCommand({
          policyType: PolicyType.DATA_PROTECTION_POLICY,
        }),
      ),
    );

    if (!response.accountPolicies) {
      throw new Error(`Undefined accountPolicies property received from DescribeAccountPolicies API.`);
    }

    return response.accountPolicies;
  }
}
