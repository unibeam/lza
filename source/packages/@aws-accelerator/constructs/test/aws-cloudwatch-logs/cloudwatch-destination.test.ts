/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { CloudWatchDestination } from '../../lib/aws-cloudwatch-logs/cloudwatch-destination';
import { snapShotTest } from '../snapshot-test';
import { describe } from '@jest/globals';

const testNamePrefix = 'Construct(CloudWatchDestination): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();
const cnStack = new cdk.Stack();

new CloudWatchDestination(stack, 'CloudWatchDestination', {
  kinesisKmsKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  kinesisStream: new cdk.aws_kinesis.Stream(stack, 'CustomStream', {}),
  organizationId: 'o-some-org-id',
  partition: 'aws',
  useExistingRoles: false,
  acceleratorPrefix: 'AWSAccelerator',
});

new CloudWatchDestination(cnStack, 'CloudWatchDestination', {
  kinesisKmsKey: new cdk.aws_kms.Key(cnStack, 'CustomKey', {}),
  kinesisStream: new cdk.aws_kinesis.Stream(cnStack, 'CustomStream', {}),
  organizationId: 'o-some-org-id',
  accountIds: ['111111111111', '222222222222'],
  partition: 'aws-cn',
  useExistingRoles: false,
  acceleratorPrefix: 'AWSAccelerator',
});

new CloudWatchDestination(stack, 'CloudWatchDestinationExistingIam', {
  kinesisKmsKey: new cdk.aws_kms.Key(stack, 'CustomKeyExistingIam', {}),
  kinesisStream: new cdk.aws_kinesis.Stream(stack, 'CustomStreamExistingIam', {}),
  organizationId: 'o-some-org-id',
  partition: 'aws',
  useExistingRoles: true,
  acceleratorPrefix: 'AWSAccelerator',
});

/**
 * CloudWatchDestination construct test
 */
describe('CloudWatchDestination', () => {
  snapShotTest(testNamePrefix, stack);
  snapShotTest(testNamePrefix, cnStack);
});
