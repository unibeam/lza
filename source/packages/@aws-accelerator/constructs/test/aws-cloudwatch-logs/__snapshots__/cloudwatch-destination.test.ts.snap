// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudWatchDestination Construct(CloudWatchDestination):  Snapshot Test 1`] = `
{
  "Conditions": {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": {
      "Fn::Or": [
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Resources": {
    "CloudWatchDestination47668810": {
      "Properties": {
        "DestinationName": "AWSAcceleratorCloudWatchToS3",
        "DestinationPolicy": {
          "Fn::Join": [
            "",
            [
              "{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":["*"]},"Action":"logs:PutSubscriptionFilter","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":destination:AWSAcceleratorCloudWatchToS3","Condition":{"StringEquals":{"aws:PrincipalOrgID":"o-some-org-id"}}}]}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "CloudWatchDestinationLogsKinesisRole54686795",
            "Arn",
          ],
        },
        "TargetArn": {
          "Fn::GetAtt": [
            "CustomStreamE8E9158E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Logs::Destination",
    },
    "CloudWatchDestinationExistingIamE1F9FD4A": {
      "Properties": {
        "DestinationName": "AWSAcceleratorCloudWatchToS3",
        "DestinationPolicy": {
          "Fn::Join": [
            "",
            [
              "{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":["*"]},"Action":"logs:PutSubscriptionFilter","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":destination:AWSAcceleratorCloudWatchToS3","Condition":{"StringEquals":{"aws:PrincipalOrgID":"o-some-org-id"}}}]}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::",
              {
                "Ref": "AWS::AccountId",
              },
              ":role/AWSAcceleratorLogReplicationRole-",
              {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "TargetArn": {
          "Fn::GetAtt": [
            "CustomStreamExistingIamFE5F8092",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Logs::Destination",
    },
    "CloudWatchDestinationLogsKinesisRole54686795": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "logs.",
                      {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "kinesis:ListShards",
                    "kinesis:PutRecord",
                    "kinesis:PutRecords",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "CustomStreamE8E9158E",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KinesisAccess",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "kms:Decrypt",
                    "kms:Encrypt",
                    "kms:GenerateDataKey",
                    "kms:ReEncryptTo",
                    "kms:GenerateDataKeyWithoutPlaintext",
                    "kms:GenerateDataKeyPairWithoutPlaintext",
                    "kms:GenerateDataKeyPair",
                    "kms:ReEncryptFrom",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "CustomKey1E6D0D07",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KmsAccess",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomKeyExistingIam1584FEB7": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomStreamE8E9158E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": {
          "Fn::If": [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            {
              "Ref": "AWS::NoValue",
            },
            {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
      },
      "Type": "AWS::Kinesis::Stream",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomStreamExistingIamFE5F8092": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": {
          "Fn::If": [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            {
              "Ref": "AWS::NoValue",
            },
            {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
      },
      "Type": "AWS::Kinesis::Stream",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`CloudWatchDestination Construct(CloudWatchDestination):  Snapshot Test 2`] = `
{
  "Conditions": {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": {
      "Fn::Or": [
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Resources": {
    "CloudWatchDestination47668810": {
      "Properties": {
        "DestinationName": "AWSAcceleratorCloudWatchToS3",
        "DestinationPolicy": {
          "Fn::Join": [
            "",
            [
              "{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":["111111111111","222222222222"]},"Action":"logs:PutSubscriptionFilter","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":destination:AWSAcceleratorCloudWatchToS3"}]}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "CloudWatchDestinationLogsKinesisRole54686795",
            "Arn",
          ],
        },
        "TargetArn": {
          "Fn::GetAtt": [
            "CustomStreamE8E9158E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Logs::Destination",
    },
    "CloudWatchDestinationLogsKinesisRole54686795": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "logs.",
                      {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "kinesis:ListShards",
                    "kinesis:PutRecord",
                    "kinesis:PutRecords",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "CustomStreamE8E9158E",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KinesisAccess",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "kms:Decrypt",
                    "kms:Encrypt",
                    "kms:GenerateDataKey",
                    "kms:ReEncryptTo",
                    "kms:GenerateDataKeyWithoutPlaintext",
                    "kms:GenerateDataKeyPairWithoutPlaintext",
                    "kms:GenerateDataKeyPair",
                    "kms:ReEncryptFrom",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::GetAtt": [
                      "CustomKey1E6D0D07",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KmsAccess",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomStreamE8E9158E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": {
          "Fn::If": [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            {
              "Ref": "AWS::NoValue",
            },
            {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
      },
      "Type": "AWS::Kinesis::Stream",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
